### DISCUSSION TITLE
Regarding auto-import feature and datetime

### DISCUSSION THREAD

**Discussion by dpetka2001** [python-lsp/python-lsp-server#540] [Q&A] [answered]
I installed `python3 -m pip install python-lsp-server` and `python3 -m pip install "python-lsp-server[rope]"` in my virtual environment to test out the `python-lsp-server`. I use Neovim and have configured `lspconfig` as follows
```lua
        pylsp = {
          settings = {
            pylsp = {
              plugins = {
                rope_autoimport = {
                  enabled = true,
                },
              },
            },
          },
        },
```
When I open a python file in my virtual environment and type `floor` then I get completion for auto-importing `math` and when typing `sleep` I get completion for auto-importing `sleep`. However, when I type `datetime` I only get completion for auto-importing `from xmlrpc.client import DateTime`. 

Am I maybe missing some configuration option to enable additional module preloading with pylsp and rope? From what I can see [here](https://github.com/python-lsp/python-lsp-server/blob/develop/pylsp/plugins/preload_imports.py), I believe that `datetime` module should be preloaded? Or maybe I have no idea what I'm talking about.

Please excuse my ignorance as I'm still new to programming and grabbed Python as my first language to start. If I do `import datetime`, then I get completion about the modules that are available. 

Appreciate any help. Thanks in advance.

**Comment by ccordoba12** [comment]
Hey @dpetka2001, thanks for reporting. I'm not sure what's happening because I don't use this functionality.

@tkrabel, any ideas about this one?

**Reply by dpetka2001** [reply]
Any advice on this? Is this something expected? Should i maybe configure something on my end?

**Comment by smheidrich** [comment, marked as answer]
# Explanation

This is a limitation of Rope: By default, it uses static analysis of Python source code to find out which names are available for import, but this can't work for C extension modules which don't have any Python source code.

Python's `datetime` module [imports its actual contents](https://github.com/python/cpython/blob/13cb8ca3da75d6fe2a1114e58c24ca681fee411f/Lib/datetime.py#L8) from a C extension module named `_datetime`, so Rope can't find out about them using this default approach.

# How to fix

Rope can be [configured](https://rope.readthedocs.io/en/latest/configuration.html) to introspect specific extension modules by actually executing them using the `extension_modules` option.

So to fix this, you can either add this to your project's `.ropeproject/config.py` file:

```python
def set_prefs(prefs):
  prefs["extension_modules"] = ["_datetime"]
```

or this to your project's `pyproject.toml` or global `pytool.toml`:

```toml
[tool.rope]
extension_modules = ['_datetime']
```

**Comment by smheidrich** [comment]
P.S. Can you please change the title to something more specific so others can find this discussion more easily? E.g.

>Regarding auto-import feature and datetime

I had the same issue and I'm sure there will be more after me.

**Reply by dpetka2001** [reply]
Wow!! Really thank you @smheidrich for the thorough explanation about what happens and how to solve it. Appreciate you taking the time and effort to take a look into this :smile:
